KAFKA-19977: Documentation section on the streams rebalance protocol (#21170)
apache/kafka: docs/streams/developer-guide/streams-rebalance-protocol.md

This approach brings Kafka Streams coordination in line with the modern broker-driven rebalance model introduced for consumers in KIP-848, providing a dedicated group type with streams-specific semantics and metadata management.

## What's Supported in This Version

The following features are available in the current release:

* **Core Streams Group Rebalance Protocol**: The `group.protocol=streams` configuration enables the dedicated streams rebalance protocol. This separates streams groups from consumer groups and provides a streams-specific group membership lifecycle and metadata management on the broker.

* **Sticky Task Assignor**: A basic task assignment strategy that minimizes task movement during rebalances is included.

* **Interactive Query Support**: IQ operations are compatible with the new streams protocol.

> * **New Admin RPC**: The StreamsGroupDescribe RPC provides streams-specific metadata separate from consumer group information, with corresponding access via the [`Admin`](/43/javadoc/org/apache/kafka/clients/admin/Admin.html).

Reviewer: ```suggestion
* **New Admin RPC**: The StreamsGroupDescribe RPC provides streams-specific metadata separate from consumer group information, with corresponding access via the [`Admin`](/{version}/javadoc/org/apache/kafka/clients/admin/Admin.html).
```

Reviewer: Cf https://github.com/apache/kafka/pull/21276

> * **New Admin RPC**: The StreamsGroupDescribe RPC provides streams-specific metadata separate from consumer group information, with corresponding access via the [`Admin`](/{version}/javadoc/org/apache/kafka/clients/admin/Admin.html).

Reviewer: ```suggestion
* **New Admin RPC**: The StreamsGroupDescribe RPC provides streams-specific metadata separate from consumer group information, with corresponding access via the [`Admin`](/{version}/javadoc/org/apache/kafka/clients/admin/Admin.html) interface.
```

> * **CLI Integration**: You can list, describe, and delete streams groups via the [kafka-streams-groups.sh](kafka-streams-group-sh.md) script.

Reviewer: ```suggestion
* **CLI Integration**: You can list, describe, and delete streams groups via the [bin/kafka-streams-groups.sh](kafka-streams-group-sh.md) script.
```

> * **Offline Migration**: After shutting down all members and waiting for their `session.timeout.ms` to expire, a classic group can be converted to a streams group and a streams group can be converted to a classic group. The only thing that will be preserved broker-side are the committed offsets of the application.

Reviewer: Why do we need to wait for `session.timeout.ms`? I guess it hold if we don't send a leave group request, what is the default, but it also possible to send leave group request explicitly now https://cwiki.apache.org/confluence/display/KAFKA/KIP-1153%3A+Refactor+Kafka+Streams+CloseOptions+to+Fluent+API+Style.

If you think we want to keep the instructions simple, I am also ok to leave it as is. Just wanted to ask.

Reviewer: > The only thing that will be preserved broker-side are the committed offsets of the application.

Could this be misleading with regard to internal topics?

# What's Not Supported in This Version

The following features are not yet available and should be avoided when using the new protocol:

* **Static Membership**: Setting a client `instance.id` will be rejected.

* **Topology Updates**: If a topology is changed significantly (e.g., by adding new source topics or changing the number of subtopologies), a new streams group must be created.

> * **High Availability Assignor**: Only the sticky assignor is supported.

Reviewer: ```suggestion
* **High Availability Assignor**: Only the sticky assignor is supported. This implies that "warmup tasks" are not supported yet.
```

* **Regular Expressions**: Pattern-based topic subscription is not supported.

* **Online Migration**: Group migration while the application is running is not available between the classic and new streams protocol.

# Why Use the Streams Rebalance Protocol?

The Streams Rebalance Protocol offers several key advantages over the classic client-driven protocol:

> * **Broker-Driven Coordination**: Centralizes task assignment logic on brokers instead of the client. This provides consistent, authoritative task assignment decisions from a single coordination point and reduces the potential for split-brain scenarios.

Reviewer: ```suggestion
* **Broker-Driven Coordination**: Centralizes task assignment logic on brokers instead of the client. This provides consistent, authoritative task assignment decisions from a single coordination point, and reduces the potential for split-brain scenarios.
```

* **Faster, More Stable Rebalances**: Reduces rebalance duration and impact by removing the global synchronization point. This minimizes application downtime during membership changes or failures.

> * **Better Observability**: Provides dedicated metrics and admin interfaces that separate streams from consumer groups, leading to clearer troubleshooting with broker-side observability. See the [streams groups metrics]({{< relref "/43/operations/monitoring#group-coordinator-monitoring" >}}) documentation for details.

Reviewer: ```suggestion
* **Better Observability**: Provides dedicated metrics and admin interfaces that separate streams from consumer groups, leading to clearer troubleshooting with broker-side observability. See the [streams groups metrics]({{< relref "/{version}/operations/monitoring#group-coordinator-monitoring" >}}) documentation for details.
```

## Enabling the Protocol

The Streams Rebalance Protocol is enabled by default on new clusters starting with Apache Kafka 4.2. Both brokers and clients must be running Apache Kafka 4.2 or later to use this protocol.

### Broker Configuration

> The protocol is enabled by default on new Apache Kafka 4.2 clusters. To enable the feature on existing clusters or to explicitly control it:

Reviewer: ```suggestion
The protocol is enabled by default on new Apache Kafka 4.2 clusters. To enable the feature on existing clusters (after upgrading to 4.2) or to explicitly control it:
```

Enable the feature:
```

> kafka-features.sh --bootstrap-server localhost:9092 upgrade --feature streams.version=1

Reviewer: ```suggestion
bin/kafka-features.sh --bootstrap-server localhost:9092 upgrade --feature streams.version=1
```

```

Disable the feature:
```

> kafka-features.sh --bootstrap-server localhost:9092 downgrade --feature streams.version=0

Reviewer: ```suggestion
bin/kafka-features.sh --bootstrap-server localhost:9092 downgrade --feature streams.version=0
```

```

### Client Configuration

In your Kafka Streams application configuration, set:
```
group.protocol=streams
```

## Configuration

### Broker Configuration

> The following broker configurations control the behavior of streams groups. For complete details, see the [broker configuration]({{< relref "/43/configuration/broker-configs" >}}) documentation.

Reviewer: ```suggestion
The following broker configurations control the behavior of streams groups. For complete details, see the [broker configuration]({{< relref "/{version}/configuration/broker-configs" >}}) documentation.
```

Reviewer: More is the same below

* [`group.coordinator.rebalance.protocols`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.coordinator.rebalance.protocols" >}}): The list of enabled rebalance protocols. `"streams"` is included in the list of protocols to enable streams groups.

> * [`group.streams.session.timeout.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.session.timeout.ms" >}}): The timeout to detect client failures when using the streams group protocol.

Reviewer: ```suggestion
* [`group.streams.session.timeout.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.session.timeout.ms" >}}): The default timeout for all streams group (if not specifically overwritten for a specific streams group) to detect client failures when using the streams group protocol.
```

> * [`group.streams.min.session.timeout.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.min.session.timeout.ms" >}}): The minimum session timeout.

Reviewer: ```suggestion
* [`group.streams.min.session.timeout.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.min.session.timeout.ms" >}}): The minimum allowed session timeout.
```

> * [`group.streams.max.session.timeout.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.max.session.timeout.ms" >}}): The maximum session timeout.

Reviewer: ```suggestion
* [`group.streams.max.session.timeout.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.max.session.timeout.ms" >}}): The maximum allowed session timeout.
```

> * [`group.streams.heartbeat.interval.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.heartbeat.interval.ms" >}}): The heartbeat interval given to the members.

Reviewer: ```suggestion
* [`group.streams.heartbeat.interval.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.heartbeat.interval.ms" >}}): The default heartbeat interval given to the members.
```

> * [`group.streams.min.heartbeat.interval.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.min.heartbeat.interval.ms" >}}): The minimum heartbeat interval.

Reviewer: ```suggestion
* [`group.streams.min.heartbeat.interval.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.min.heartbeat.interval.ms" >}}): The minimum allowed heartbeat interval.
```

> * [`group.streams.max.heartbeat.interval.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.max.heartbeat.interval.ms" >}}): The maximum heartbeat interval.

Reviewer: ```suggestion
* [`group.streams.max.heartbeat.interval.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.max.heartbeat.interval.ms" >}}): The maximum allowed heartbeat interval.
```

* [`group.streams.max.size`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.max.size" >}}): The maximum number of streams clients that a single streams group can accommodate.

> * [`group.streams.num.standby.replicas`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.num.standby.replicas" >}}): The number of standby replicas for each task.

Reviewer: ```suggestion
* [`group.streams.num.standby.replicas`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.num.standby.replicas" >}}): The default number of standby replicas for each task.
```

* [`group.streams.max.standby.replicas`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.max.standby.replicas" >}}): Maximum for dynamic configurations of the standby replica configuration.

> * [`group.streams.initial.rebalance.delay.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.initial.rebalance.delay.ms" >}}): The first rebalance of a group is delayed by this amount to allow more members to join the group.

Reviewer: ```suggestion
* [`group.streams.initial.rebalance.delay.ms`]({{< relref "/43/configuration/broker-configs#brokerconfigs_group.streams.initial.rebalance.delay.ms" >}}): The first rebalance of a new (ie, previously empty) group is delayed by this amount to allow more members to join the group.
```

Reviewer: Again, it's a default but can be overwritten on a per-group basis.

### Group Configuration

> Configurations for the resource type `GROUP` are available in `DescribeConfigs` and `IncrementalAlterConfigs` to override the default broker configurations dynamically for specific groups. These can be set using the [`Admin`](/43/javadoc/org/apache/kafka/clients/admin/Admin.html) Java interface or the `kafka-configs.sh` utility.

Reviewer: ```suggestion
Configurations for the resource type `GROUP` are available in `DescribeConfigs` and `IncrementalAlterConfigs` to override the default broker configurations dynamically for specific groups. These can be set using the [`Admin`](/{version}/javadoc/org/apache/kafka/clients/admin/Admin.html) Java interface or the `bin/kafka-configs.sh` utility.
```

> For complete details, see the [group configuration]({{< relref "/43/configuration/group-configs" >}}) documentation.

Reviewer: ```suggestion
For complete details, see the [group configuration]({{< relref "/{version}/configuration/group-configs" >}}) documentation.
```

The following group-level configurations are available for streams groups:

> * [`streams.session.timeout.ms`]({{< relref "/43/configuration/group-configs#groupconfigs_streams.session.timeout.ms" >}}): The timeout to detect client failures when using the streams group protocol.

Reviewer: ```suggestion
* [`streams.session.timeout.ms`]({{< relref "/{version}/configuration/group-configs#groupconfigs_streams.session.timeout.ms" >}}): The timeout to detect client failures when using the streams group protocol.
```

Reviewer: More of the same below

* [`streams.heartbeat.interval.ms`]({{< relref "/43/configuration/group-configs#groupconfigs_streams.heartbeat.interval.ms" >}}): The heartbeat interval given to the members.
* [`streams.num.standby.replicas`]({{< relref "/43/configuration/group-configs#groupconfigs_streams.num.standby.replicas" >}}): The number of standby replicas for each task.
* [`streams.initial.rebalance.delay.ms`]({{< relref "/43/configuration/group-configs#groupconfigs_streams.initial.rebalance.delay.ms" >}}): The first rebalance of a group is delayed by this amount to allow more members to join the group.

#### Example: Setting Group-Level Configuration
```

> kafka-configs.sh --bootstrap-server localhost:9092 \

Reviewer: ```suggestion
bin/kafka-configs.sh --bootstrap-server localhost:9092 \
```

  --alter --entity-type groups --entity-name wordcount \
  --add-config streams.num.standby.replicas=1
```

> **Note:** In the streams rebalance protocol, `session.timeout.ms`, `heartbeat.interval.ms` and `num.standby.replicas` are group-level configurations, which are ignored when they are set on the client side. Use the `kafka-configs.sh` tool to set these configurations as shown above.

Reviewer: ```suggestion
**Note:** In the streams rebalance protocol, `session.timeout.ms`, `heartbeat.interval.ms` and `num.standby.replicas` are group-level configurations, which are ignored when they are set on the client side. Use the `bin/kafka-configs.sh` tool to set these configurations as shown above.
```

### Streams Configuration

> For complete details on all Kafka Streams configurations, see the [streams configuration]({{< relref "/43/configuration/kafka-streams-configs" >}}) documentation.

Reviewer: ```suggestion
For complete details on all Kafka Streams configurations, see the [streams configuration]({{< relref "/{version}/configuration/kafka-streams-configs" >}}) documentation.
```

The following client configuration enables the streams rebalance protocol:

> * [`group.protocol`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_group.protocol" >}}): A flag which indicates if the streams rebalance protocol should be used. Set to `streams` to enable (default is `classic`).

Reviewer: ```suggestion
* [`group.protocol`]({{< relref "/{version}/configuration/kafka-streams-configs#streamsconfigs_group.protocol" >}}): A flag which indicates if the streams rebalance protocol should be used. Set to `streams` to enable (default is `classic`).
```

#### Ignored Configurations

The following configurations are ignored when the streams rebalance protocol is enabled:

> * [`acceptable.recovery.lag`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_acceptable.recovery.lag" >}})

Reviewer: ```suggestion
* [`acceptable.recovery.lag`]({{< relref "/{version}/configuration/kafka-streams-configs#streamsconfigs_acceptable.recovery.lag" >}})
```

Reviewer: More of the same below.

* [`probing.rebalance.interval.ms`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_probing.rebalance.interval.ms" >}})
* [`rack.aware.assignment.tags`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_rack.aware.assignment.tags" >}})
* [`rack.aware.assignment.strategy`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_rack.aware.assignment.strategy" >}})
* [`rack.aware.assignment.traffic_cost`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_rack.aware.assignment.traffic_cost" >}})
* [`rack.aware.assignment.non_overlap_cost`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_rack.aware.assignment.non_overlap_cost" >}})
* [`task.assignor.class`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_task.assignor.class" >}})
* [`session.timeout.ms`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_session.timeout.ms" >}}) (use group-level configuration instead)
* [`heartbeat.interval.ms`]({{< relref "/43/configuration/kafka-streams-configs#streamsconfigs_heartbeat.interval.ms" >}}) (use group-level configuration instead)

## Administration

### Admin API

> Use the "streams groups" methods of the [`Admin`](/43/javadoc/org/apache/kafka/clients/admin/Admin.html) interface to manage streams groups programmatically. These APIs are mostly backed by the same implementations as the consumer group API.

Reviewer: ```suggestion
Use the "streams groups" methods of the [`Admin`](/{version}/javadoc/org/apache/kafka/clients/admin/Admin.html) interface to manage streams groups programmatically. These APIs are mostly backed by the same implementations as the consumer group API.
```

The main differences from consumer group APIs are:

* The `describeStreamsGroups` uses the DescribeStreamsGroup RPC and contains different information than consumer groups.
* A streams group has an extra state - `NOT_READY` - and no legacy states from the classic protocol.
* `removeMembersFromConsumerGroup` will not have a corresponding API in this version, as it uses the LeaveGroup RPC for classic consumer groups, which is not available for KIP-848-style groups.

### kafka-streams-groups.sh

> A new tool called [kafka-streams-groups.sh](kafka-streams-group-sh.md) is added for working with streams groups. It replaces `kafka-streams-application-reset` for streams groups and can be used to list, describe, and delete streams groups. See the [kafka-streams-groups.sh documentation](kafka-streams-group-sh.md) for detailed usage information.

Reviewer: ```suggestion
A new tool called [bin/kafka-streams-groups.sh](kafka-streams-group-sh.md) is added for working with streams groups. It replaces `bin/kafka-streams-application-reset.sh` for streams groups and can be used to list, describe, and delete streams groups. See the [bin/kafka-streams-groups.sh documentation](kafka-streams-group-sh.md) for detailed usage information.
```

> A new tool called `bin/kafka-streams-groups.sh` is added for working with streams groups. It replaces `kafka-streams-application-reset` for streams groups and can be used to list, describe, and delete streams groups. See the [kafka-streams-groups.sh documentation](/{version}/streams/developer-guide/kafka-streams-group-sh/) for detailed usage information.

Reviewer: ```suggestion
A new tool called `bin/kafka-streams-groups.sh` is added for working with streams groups. It replaces `bin/kafka-streams-application-reset.sh` for streams groups and can be used to list, describe, and delete streams groups. See the [kafka-streams-groups.sh documentation](/{version}/streams/developer-guide/kafka-streams-group-sh/) for detailed usage information.
```

* Check that all internal topics exist with the right configuration.

If any source topics or internal topics are missing, the group enters a state `NOT_READY`. In `NOT_READY`, all heartbeats will be handled as usual (so they typically should not fail), but in the heartbeat response, the status will indicate which kind of problem exists. All members will get an empty assignment when the group is in `NOT_READY` state.

### Centralized Assignment Configuration

Core assignment options are configured centrally on the broker, without relying on each client's configuration. This allows tuning a streams group without redeploying the streams application. The core assignment option introduced on the broker-side is `num.standby.replicas`. This can be configured both globally on the broker and dynamically for specific streams groups through the `IncrementalAlterConfigs` and `DescribeConfigs` RPCs.

The last used assignment configuration is stored in the group metadata on the broker. This way, if an assignment configuration is dynamically changed, reassignment can be triggered immediately.

## Monitoring and Metrics

> The existing group metrics are extended to differentiate between streams groups and consumer groups and account for streams group states. For complete details, see the [streams groups metrics]({{< relref "/43/operations/monitoring#group-coordinator-monitoring" >}}) documentation.

Reviewer: ```suggestion
The existing group metrics are extended to differentiate between streams groups and consumer groups and account for streams group states. For complete details, see the [streams groups metrics]({{< relref "/{version}/operations/monitoring#group-coordinator-monitoring" >}}) documentation.
```

### Group Count by Protocol

Number of groups based on type of protocol, where the list of protocols is extended by the `protocol=streams` variation:
```
kafka.server:type=group-coordinator-metrics,name=group-count,protocol={consumer|classic|streams}
```

### Streams Group Count by State

Number of streams groups based on state:
```
kafka.server:type=group-coordinator-metrics,name=streams-group-count,state={empty|not_ready|assigning|reconciling|stable|dead}
```

### Streams Group Rebalances

Streams group rebalances sensor:
```
kafka.server:type=group-coordinator-metrics,name=streams-group-rebalance-rate

kafka.server:type=group-coordinator-metrics,name=streams-group-rebalance-count
```

## Migration from Classic Protocol

Currently, only offline migration is supported. To migrate a Kafka Streams application from the classic protocol to the streams rebalance protocol:

1. Shut down all application instances.

> 2. Wait for the `session.timeout.ms` to expire so the group becomes empty.

Reviewer: Similar to above.

3. Update the application configuration to set `group.protocol=streams`.
4. Restart the application instances.

> The only broker-side data that will be preserved are the committed offsets of the application. All other group metadata will be recreated when the application starts with the new protocol.

Reviewer: As above. Misleading with regard to internal topics?
